Lib/functools.py

```diff
@@ -928,27 +928,41 @@ class singledispatchmethod:
     """
 
     def __init__(self, func):
+        import weakref # lazy import
         if not callable(func) and not hasattr(func, "__get__"):
             raise TypeError(f"{func!r} is not callable or a descriptor")
 
         self.dispatcher = singledispatch(func)
         self.func = func
+        self._method_cache = weakref.WeakKeyDictionary()
+        self._all_weakrefable_instances = True
 
     def register(self, cls, method=None):
         """generic_method.register(cls, func) -> func
 
         Registers a new implementation for the given *cls* on a *generic_method*.
         """
+        self._method_cache.clear()
```

Comment:
Why do we need to do the `cache_clear()` here? Is there a test you can write that would fail if we didn't have the `cache_clear()` here?

Reply:
(To be clear, this wasn't me saying "I demand you remove this" -- it was a genuine question :)

Reply:
Adding the `_method_cache.clear()` was a mistake in reasoning from my side. I cannot think of any cases where we need the `clear()`